### PR TITLE
feat: Add Fuzz and Fuzz CLI

### PR DESCRIPTION
**Pull request by rlaau**
### feat: Add Fuzz and Fuzz CLI  

Go has a native fuzzer written in Go, but Gno does not yet have such a fuzzer.  
I thought it would be useful to have a simple fuzzer that can be imported and used via the `testing` library within Gno files.  

Therefore, I have added fuzzing-related functionalities to `stdlibs/testing` and `stdlibs/testing/fuzzing`.  
Users can now import this fuzzer from the `testing` library and use it just like Go’s native fuzzer, but as a Gno-native fuzzer.  

---

### Goal  

1. Provide a Gno fuzzer with the same interface as Go’s fuzzer for ease of use.  
2. Ensure that the Gno fuzzer runs within the GnoVM execution logic, allowing it to detect errors specific to Gno programs.  
3. Maintain a certain level of performance for fuzzing.  

---

### Changes  

- **Fix:** Updated  
  - `gnovm/cmd/gno/test.go`  
  - `gnovm/pkg/test/test.go`  
  - `gnovm/stdlibs/testing/testing.gno`  
  - Enabled calling the fuzzer via CLI.  

- **Add:** `gnovm/stdlibs/testing/fuzz.gno` – Manages the interface for the Gno fuzzer.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/fuzz_hasher.gno` – Assigns and manages IDs for test coverage using `FuzzHasher`.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/fuzz_logger.gno` – Manages error inputs and logs via `FuzzLogger`.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/fuzz_manager.gno` – Efficiently manages seed inputs and coverage (hash numbers) priorities via `FuzzManager`.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/get_coverage.gno` – Executes input values to collect coverage data.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/mutator.gno` – Manages the seed values to be mutated and generates the next input from those seeds.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/parser.gno`, `gnovm/stdlibs/testing/fuzzing/parser_for_not_sb.gno` – Analyzes input arguments to help the mutator perform more effective transformations.  
- **Add:** `gnovm/stdlibs/testing/fuzzing/random.gno` – Generates PCG random numbers for fuzzing.  

---

### Algorithm  

1. The user provides initial seeds via `f.Add`. All initial seeds are executed, and based on the results:  
   - They are coordinated by `f.manager`.  
   - `f.mutator` stores the seeds along with analyzed execution results.  

2. When `f.Fuzz` receives the target program to test, the following loop begins:  

   1. `f.manager` selects the highest-priority coverage (the least executed one) and extracts a seed from its queue, passing it to `f.mutator`.  
   2. `f.mutator` analyzes the seed and selects an appropriate mutation strategy to modify it.  
   3. The fuzzer and `f.hasher` execute the mutated seed and classify the execution results (assigning a hash number to coverage).  
   4. The results are processed:  
      - `f.manager` coordinates the seed based on execution results.  
      - `f.mutator` updates mutation strategies according to coverage and seed data.  
   5. If a failure occurs, the fuzzer stops execution, and `f.logger` reports the error.  
   6. Or, the fuzzer terminates when the predefined number of iterations is completed.
  
gfalg

---

### Usage  

1. Currently, coverage measurement for arbitrary functions is not implemented. The following example demonstrates coverage measurement for a specific function where coverage tracking is possible.  
2. The CLI command allows usage similar to Go’s fuzzer:  

   ```sh
   gno test [file name] -fuzz=[function prefix] -v -i=[iteration count]
3. Other functionalities follow the usage described in [Go’s fuzz documentation](https://go.dev/doc/security/fuzz/).
4. However, there are a few differences:
- The anonymous function passed to f.Fuzz must follow the format:
- func(t *testing.T, args …interface{})
- The fuzzing termination condition is iteration-based rather than time-based.
5. Otherwise, usage is identical.
```go
package testing_test

import (
	"testing"
	"testing/fuzzing"
	"unicode/utf8"
)

/// TODO: Once coverage is fully implemented, test whether the fuzzer can detect
//       issues such as HTTP request failures or compilation errors.

// FuzzEdgeCase tests whether the fuzzer can effectively generate edge cases.
// It evaluates the ability to produce edge inputs in a structured manner.
func FuzzEdgeCase(f *testing.F) {
	f.Add("apple hello", int(42131231230))
	f.Add("rainy day", int(98401132231331))
	f.Add("winter comes", int(12349123123))
	f.Fuzz(func(t *testing.T, orig ...interface{}) {
		v, ok := orig[0].(string)
		if !ok {
			panic("dont match")
		}
		i, ok2 := orig[1].(int)
		if !ok2 {
			panic("dont match")
		}
		rev := fuzzing.Reverse(v)
		doubleRev := fuzzing.Reverse(rev)
		if v != doubleRev && i > 300 && i < 500 {
			t.Errorf("Before: %q, after: %q", orig, doubleRev)
		}
		if utf8.ValidString(v) && !utf8.ValidString(rev) && i > 300 && i < 1000 {
			t.Errorf("Reverse produced invalid UTF-8 string %q", rev)
		}
	})
}

// FuzzSymbolicPath tests whether the fuzzer can explore symbolic execution paths.
// It assesses the ability to traverse deep path depths effectively.
func FuzzSymbolicPath(f *testing.F) {
	f.Add("")
	f.Fuzz(func(t *testing.T, orig ...interface{}) {
		s, ok := orig[0].(string)
		if !ok {
			panic("dont match")
		}
		if len(s) > 0 && s[0] == 'b' {
			if len(s) > 1 && s[1] == 'a' {
				if len(s) > 2 && s[2] == 'd' {
					if len(s) > 3 && s[3] == '!' {
						panic("crash triggered")
					}
				}
			}
		}
	})
}
```
gf

---

### Notes
- Since coverage tracking has not yet been fully implemented, this fuzzer cannot be used on all code immediately.
- The fuzzer was designed with future extensibility in mind and requires ongoing improvements, such as:
-- Refining mutation strategies
-- Improving management logic
-- Optimizing priority selection
-- Enhancing sampling techniques
- However, to prevent the PR from becoming too large, I have submitted only the minimal functionality for now.

---

### Future Enhancements
- TODOs:
-- Additional input minimization
-- Refining mutation and strategy update methods
-- Pattern-based analysis
-- Implement full coverage tracking
- Apply sampling techniques for seed selection (e.g., Thompson sampling)

- Set an execution limit to stop exploration beyond a threshold and reset strategies

- Improve transformation and priority determination using Sonar-like distance metrics

- Extract execution logic separately and process it in parallel using multiprocessing, assigning each process its own GnoVM instance for testing

### CODE REVIEW

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: rlaau/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 If
    ├── 🔴 Condition
    │   └── 🔴 Or
    │       ├── 🔴 At least 1 user(s) of the organization reviewed the pull request (with state "APPROVED")
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🔴 Else
        └── 🔴 And
            ├── 🟢 This label is applied to pull request: review/triage-pending
            └── 🔴 On no pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

File: gnovm/cmd/gno/test.go [hunk 1]
```diff
@@ -218,9 +236,15 @@ func execTest(cfg *testCfg, args []string, io commands.IO) error {
 			err = test.Test(memPkg, pkg.Dir, opts)
 		})
 
+		if cfg.fuzzName != "" {
+			if testErrCount > 0 || buildErrCount > 0 {
+				return fmt.Errorf("   --- %d build errors, %d test errors", buildErrCount, testErrCount)
+			} else {
+				return nil
+			}
+		}
```
Comment (notJoon) [inline gnovm/cmd/gno/test.go:245]:
```suggestion
		if cfg.fuzzName != "" {
			if testErrCount > 0 || buildErrCount > 0 {
				return fmt.Errorf("   --- %d build errors, %d test errors", buildErrCount, testErrCount)
			}
			return nil
		}
```

**Review by notJoon — CHANGES_REQUESTED** [review]
I've left some comments. Also, please translate all comments to English.

File: gnovm/stdlibs/testing/fuzzing/parser.gno [hunk 2]
```diff
@@ -0,0 +1,1818 @@
...
+					xCharBytes := lx.next()
+					xChar, _ := utf8.DecodeRune(xCharBytes)
+					hexDigits := ""
+					for !lx.eof() {
+						pb = lx.peek()
+						if pb == nil {
+							break
+						}
+						r, _ = utf8.DecodeRune(pb)
+						if (r >= '0' && r <= '9') || (r >= 'a' && r <= 'f') || (r >= 'A' && r <= 'F') {
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno]: resolved it. #https://github.com/gnolang/gno/pull/3769/commits/08422435913a3e3db9c41c38f0c76f38e89fc40f

[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno]: resolved it. https://github.com/gnolang/gno/pull/3769/commits/08422435913a3e3db9c41c38f0c76f38e89fc40f

File: gnovm/stdlibs/testing/fuzzing/parser.gno [hunk 3]
```diff
@@ -0,0 +1,1818 @@
...
+		// consumes '0' first, then consumes 'x' or 'X'.
+		if r == '0' {
+			if lx.pos+1 < len(lx.input) {
+				nb := lx.input[lx.pos+1:]
+				nr, _ := utf8.DecodeRune(nb)
+				if nr == 'x' || nr == 'X' {
+					lx.next()
+					xCharBytes := lx.next()
+					xChar, _ := utf8.DecodeRune(xCharBytes)
+					hexDigits := ""
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno]: I have modified the string concatenation (+ operator) used inside the loop to use strings.Builder. For string concatenation outside the loop, I left them as they are because using strings.Builder might not significantly improve performance. However, if I am mistaken, please let me know.
https://github.com/gnolang/gno/pull/3769/commits/0a3e3e9d102d486e846325e2c39866ed5e6cb64d

File: gnovm/stdlibs/testing/fuzzing/parser.gno [hunk 4]
```diff
@@ -0,0 +1,1818 @@
...
+				break
+			}
+			sb.WriteRune(r)
+			lx.next()
+		}
+		return &LetterEnt{Data: []byte(sb.String())}
+	}
+
+	// 5) Identifies multi-character markers.
+	sorted := sortMarksByLengthDesc(parserMark)
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno]: i fixed it. https://github.com/gnolang/gno/pull/3769/commits/fd2eb6f023bdf5d8a117b310888c2995fb15830e

File: gnovm/stdlibs/testing/fuzzing/parser_for_not_sb.gno [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	return uint64(v)
+}
+
+// factorize attempts to find prime factors of n within 500 iterations.
+// If successful, it returns the factors and true; otherwise, nil and false.
+func factorize(n uint64) ([]uint64, bool) {
+	var factors []uint64
+	iterations := 0
+
+	trialBoundary := 500
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser_for_not_sb.gno:96]: added annotation about that
In summary, 500 is a heuristic value chosen to balance the success rate of prime factorization for numbers above 10⁶ while ensuring the process does not take too long.
https://github.com/gnolang/gno/pull/3769/commits/f2bfebc0894a9dbd1ae4ddea2fa3f07abf0cc73a

File: gnovm/stdlibs/testing/fuzzing/parser.gno [hunk 6]
```diff
@@ -0,0 +1,1818 @@
+package fuzzing
+
+import (
+	"bytes"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno:5]: I have a few questions regarding this comment:

If the ufmt package is imported by the testing/fuzzing library in its current state, it will cause an import cycle error. Of course, this can be resolved by changing the package of ufmt_test.gno.

Separately, I need to understand the meaning of "Currently fmt is only supported in the test environment" more clearly. All the code within the testing and testing/fuzzing packages is designed to work only in the test environment. Moreover, I noticed that the package_sort file in gno allows importing fmt within the testing package, and the testing.T struct also uses the fmt package. From this, I concluded that gno permits the use of packages like fmt and os in the test environment.

However, I'm not entirely sure if I correctly understand what exactly is meant by this "test environment." I would appreciate it if you could clarify this for me.

File: gnovm/stdlibs/testing/fuzzing/parser.gno [hunk 7]
```diff
@@ -0,0 +1,1818 @@
...
+	"unicode/utf8"
+)
+
+// Ent represents a structure used for analyzing string and []byte values.
+// Strings are parsed into a tree structure using Ent as an entity.
+type Ent interface {
+	// setDepth assigns a depth value to the entity from a tree perspective.
+	setDepth(d int)
+	// getDepth retrieves the depth value of the entity.
+	getDepth() int
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno]: resolved it. https://github.com/gnolang/gno/pull/3769/commits/7e4ed154eab93b1edfb0a86f24facbb93248cfbf

File: gnovm/stdlibs/testing/fuzzing/parser.gno (hunk 6 above)
[the remark this answers is not in the document]
Comment (notJoon) [reply gnovm/stdlibs/testing/fuzzing/parser.gno:5]: It would be good to refer to this part. The `os` and `fmt` packages are features temporarily injected into the test environment. As noted in the comments, they are unstable and may be deleted at some point, so please do not use them in feature implementation.

https://github.com/gnolang/gno/blob/955e1cadfe47b0a299382019681182cfc6182709/gnovm/pkg/test/imports.go#L95-L128

File: gnovm/stdlibs/testing/fuzzing/mutator.gno [hunk 8]
```diff
@@ -0,0 +1,3204 @@
...
+
+	exponent := (bits >> 23) & 0xFF
+	mantissa := bits & 0x7FFFFF
+	sign := bits & 0x80000000
+	t := uint32(UnixNano())
+	manshift := 1 + (t % 7)
+
+	var shift int8
+	if exponent <= 1 {
+		shift = int8(1 + int(mantissa%2))
```
Comment (notJoon) [inline gnovm/stdlibs/testing/fuzzing/mutator.gno:3158]: Changing `mantissa%2` to `mantissa&1` allows for more efficient processing because it only checks the LSB (Least Significant Bit) instead of performing an actual division operation.

You can also apply this method in `randomFloat64` too.

File: gnovm/stdlibs/testing/fuzzing/mutator.gno [hunk 9]
```diff
@@ -0,0 +1,3204 @@
...
+}
+
+func randomBool() bool {
+	return UniformRandomBool(0.5)
+}
+
+func randomFloat32(a float32) float32 {
+	bits := math.Float32bits(a)
+
+	exponent := (bits >> 23) & 0xFF
```
Comment (notJoon) [inline gnovm/stdlibs/testing/fuzzing/mutator.gno:3150]: You can reduce unnecessary type conversions by setting `exponent` to `int32` type. In this case `shift` should also be changed to `int32` type.

File: gnovm/stdlibs/testing/fuzzing/random.gno
Comment (notJoon) [inline gnovm/stdlibs/testing/fuzzing/random.gno:1]: It would be good to add a comment explaining that the pcg from stdlib couldn't be used due to an import cycle, which is why it was separately added to the testing package.

File: gnovm/stdlibs/testing/fuzzing/mutator.gno
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/mutator.gno:1]: I have made the modifications. I considered applying this refactoring to the mutator logic for numeric types as well, but it seemed unnecessary in that case. So, for now, I have only proceeded with refactoring for the Ent type.

File: gnovm/stdlibs/testing/fuzzing/parser.gno [hunk 10]
```diff
@@ -0,0 +1,1818 @@
...
+	// getStrategy2count retrieves the overall scores for each mutation strategy.
+	getStrategy2count() map[MutationCode]uint64
+
+	// Caching methods for mutation strategies
+	// getStrategyCache retrieves an array of mutation strategies along with
+	// their cumulative scores. The mutator uses this to select a strategy based
+	// on its score distribution.
+	getStrategyCache() ([]MutationCode, []uint64)
+	// updateStrategyCache updates the cumulative score array for mutation strategies.
+	updateStrategyCache()
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/parser.gno:52]: Considering that the keys themselves are sequential integers starting from 0, I replaced them with a simple array. Instead of having the Ent structure sort both keys and cumulative sums into separate arrays, I created a cumulative sum array indexed directly by the keys. As a result, the sorting function was no longer needed, so I removed it in subsequent commits after this one.
https://github.com/gnolang/gno/pull/3769/commits/6e6a7659e1fdbf8320f37e9cba752fb6b9262038

File: gnovm/cmd/gno/test.go (hunk 1 above)
Comment (rlaau) [reply gnovm/cmd/gno/test.go:245]: received the commit

File: gnovm/stdlibs/testing/fuzzing/mutator.gno [hunk 11]
```diff
@@ -0,0 +1,3213 @@
...
+			IsStringOrByteArray:  true,
+			EntId:                entID,
+			UsedStrategyMap:      usedStrategyMap,
+			UsedCodesForNotSbMap: nil,
+		}
+	case bool:
+		result, usedStrategyMap := MutateNotSbEnt2interface(ms.NotSbEntsDict[argIdx], strength)
+		boolResult, ok := result.(bool)
+		if !ok {
+			panic("logical Error. If it occred, Needs to fix")
```
[the remark this answers is not in the document]
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/mutator.gno]: https://github.com/gnolang/gno/pull/3769/commits/a001b82fb66b7d0ce5b28e40831d8d001c35dcb8
resolved it.

File: gnovm/stdlibs/testing/fuzzing/mutator.gno (hunk 8 above)
Comment (rlaau) [reply gnovm/stdlibs/testing/fuzzing/mutator.gno]: https://github.com/gnolang/gno/pull/3769/commits/a001b82fb66b7d0ce5b28e40831d8d001c35dcb8
resolved it

**Review by notJoon — APPROVED** [review]
LGTM

**Comment by kristovatlas** [comment]
@rlaau can I get in touch with you via email please? :)

**Comment by rlaau** [comment]
> @rlaau can I get in touch with you via email please? :)

Sure, feel free to contact  me at rlaaudgjs5638@naver.com. I check my email every weekend.

**Comment by nemanjantic** [comment]
Closing as outdated. Feel free to reopen if it's still relevant.
